src/runtime/gc_conservative.go

```diff
@@ -340,6 +340,30 @@ func alloc(size uintptr) unsafe.Pointer {
 	}
 }
 
+//go:noinline
+func realloc(ptr unsafe.Pointer, size uintptr) unsafe.Pointer {
+	newAlloc := alloc(size)
+	if ptr == nil {
+		return newAlloc
+	}
+
+	ptrAddress := uintptr(ptr)
+	endOfTailAddress := blockFromAddr(ptrAddress).findNext().address()
+
+	//this might be a few bytes longer than the original size of
+	//ptr, because we align to full blocks of size bytesPerBlock
+	ptrSize := endOfTailAddress - ptrAddress - 1
```

Comment:
This looks like an off-by-one.

```suggestion
	ptrSize := endOfTailAddress - ptrAddress
```

Also, perhaps the name `oldSize` or `objectSize` is a bit clearer than `ptrSize`? It's not the size of the pointer, it's the size of the object that the pointer points to.

Reply:
Agree to rename it. Not sure about the off-by-one. Afaik block.findNext() returns the next head/free block, which is already outside of the previous allocation. Therefore - 1 was supposed to calculate till the last byte of the allocation.

Reply:
> Not sure about the off-by-one. Afaik block.findNext() returns the next head/free block, which is already outside of the previous allocation.

That's correct.

> Therefore - 1 was supposed to calculate till the last byte of the allocation.

No. That's an off-by-one.
Imagine you have an allocation of two blocks, which will be 32 bytes (assuming a 32-bit system). Let's say it is located at address 1024. Two blocks further will be 1024 + 32 = 1056. If you have only those two addresses and want to calculate the size of the first one, you can subtract the two: 1056 - 1024 = 32.
You may be confused by the fact that addresses start at 0. So the block at address 1024 can be indexed from byte 0 (address 1024) up to and including byte 31 (address 1024+31=1055). The full size is 32 bytes, but offsets 0..31 are inside the allocation.

Reply:
Thank you for the explanation! This helped a lot!